sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/PartitionBasedLoadBalancer.java

```diff
@@ -0,0 +1,331 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.messaging.eventhubs.implementation;
+
+import static java.util.stream.Collectors.mapping;
+import static java.util.stream.Collectors.toList;
+
+import com.azure.core.implementation.util.ImplUtils;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.messaging.eventhubs.EventHubAsyncClient;
+import com.azure.messaging.eventhubs.EventHubAsyncConsumer;
+import com.azure.messaging.eventhubs.EventProcessor;
+import com.azure.messaging.eventhubs.PartitionManager;
+import com.azure.messaging.eventhubs.models.PartitionOwnership;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import reactor.core.publisher.Mono;
+import reactor.util.function.Tuple2;
+
+/**
+ * This class is responsible for balancing the load of processing events from all partitions of an Event Hub by
+ * distributing the number of partitions uniformly among all the  active {@link EventProcessor}s.
+ * <p>
+ * This load balancer will retrieve partition ownership details from the {@link PartitionManager} to find the number of
+ * active {@link EventProcessor}s. It uses the last modified time to decide if an EventProcessor is active. If a
+ * partition ownership entry has not be updated for a specified duration of time, the owner of that partition is
+ * considered inactive and the partition is available for other EventProcessors to own.
+ * </p>
+ * More details can be found <a href="https://gist.github.com/srnagar/7ef8566cfef0673288275c450dc99590">here</a>.
```

Comment:
It may be good for this to live in a more official place, like one of our READMEs. (Tho, at this point, our README needs a table of contents. The Azure identity README has a TOC.)